torch/nn/modules/module.py

```diff
@@ -593,79 +617,103 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
+        local_state = {k: v.data for k, v in local_name_params if v is not None}
+
+        for name, param in local_state.items():
+            key = prefix + name
+            if key in state_dict:
+                input_param = state_dict[key]
+                if isinstance(input_param, Parameter):
                     # backwards compatibility for serialized parameters
-                    param = param.data
+                    input_param = input_param.data
                 try:
-                    own_state[name].copy_(param)
+                    param.copy_(input_param)
                 except Exception:
-                    raise RuntimeError('While copying the parameter named {}, '
+                    raise RuntimeError('While copying the parameter named "{}", '
                                        'whose dimensions in the model are {} and '
                                        'whose dimensions in the checkpoint are {}.'
-                                       .format(name, own_state[name].size(), param.size()))
+                                       .format(key, param.size(), input_param.size()))
             elif strict:
-                unexpected.append(name)
+                missing_keys.append(key)
+
+        if strict:
+            for key, input_param in state_dict.items():
+                if key.startswith(prefix):
+                    input_name = key[len(prefix):]
+                    if '.' not in input_name and input_name not in local_state:
+                        unexpected_keys.append(key)
+
+    def load_state_dict(self, state_dict, strict=True):
+        r"""Copies parameters and buffers from :attr:`state_dict` into
+        this module and its descendants. If :attr:`strict` is ``True``, then
+        the keys of :attr:`state_dict` must exactly match the keys returned
+        by this module's :meth:`~torch.nn.Module.state_dict` function.
+
+        Arguments:
+            state_dict (dict): a dict containing parameters and
+                persistent buffers.
+            strict (bool, optional): whether to strictly enforce that the keys
+                in :attr:`state_dict` match the keys returned by this module's
+                :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
+        """
+        missing_keys = []
+        unexpected_keys = []
+
+        # copy state_dict so _load_from_state_dict can modify it
+        metadata = state_dict._metadata if hasattr(state_dict, '_metadata') else None
+        state_dict = state_dict.copy()
+        if metadata is not None:
+            state_dict._metadata = metadata
+
+        def load(module, prefix=''):
+            module._load_from_state_dict(state_dict, prefix, strict, missing_keys, unexpected_keys)
+            for name, child in module._modules.items():
+                if child is not None:
+                    load(child, prefix + name + '.')
```

Comment:
We should maintain a per-module memo here. Otherwise you can end up with multiple modules getting called multiple times, and because e.g. `missing_keys` is not a `set`, then some keys can be reported as missing multiple times. We don't have an invariant saying that modules form a tree. They can be a general DAG or regular graph.

Reply:
Right.. but if it is a dag like below:
```
  A 
 /   \
B   C
 \   /
  D

all arrows from top to bottom
```
Then `D`'s param will still occur twice in `A.state_dict()` each time with different keys, right?